File: sdk/go/pkg/artifact/nodejs.go
package artifact

import (
	"fmt"

	api "github.com/ALT-F4-LLC/vorpal/sdk/go/pkg/api/artifact"
	"github.com/ALT-F4-LLC/vorpal/sdk/go/pkg/config"
)

func NodeJS(context *config.ConfigContext) (*string, error) {
	name := "nodejs"

	system := context.GetTargetStr()

	var sourceTarget string
	switch system {
	case "aarch64-darwin":
		sourceTarget = "darwin-arm64"
	case "aarch64-linux":
		sourceTarget = "linux-arm64"
	case "x86_64-darwin":
		sourceTarget = "darwin-x64"
	case "x86_64-linux":
		sourceTarget = "linux-x64"
	default:
		return nil, fmt.Errorf("unsupported %s system: %s", name, system)
	}

	sourceVersion := "22.22.0"
	sourcePath := fmt.Sprintf(
		"https://sdk.vorpal.build/source/node-v%s-%s.tar.gz",
		sourceVersion, sourceTarget,
	)

	source := NewArtifactSource(name, sourcePath).Build()

	stepScript := fmt.Sprintf(
		"cp -pr \"./source/%s/node-v%s-%s/.\" \"$VORPAL_OUTPUT\"",
		name, sourceVersion, sourceTarget,
	)

	step, err := Shell(context, []*string{}, []string{}, stepScript, []*api.ArtifactStepSecret{})
	if err != nil {
		return nil, err
	}

	return NewArtifact(name, []*api.ArtifactStep{step}, config.SYSTEMS).
		WithAliases([]string{fmt.Sprintf("%s:%s", name, sourceVersion)}).
		WithSources([]*api.ArtifactSource{&source}).
		Build(context)
}

File: sdk/python/src/vorpal_sdk/artifact/nodejs.py
"""Node.js runtime artifact (official binary distribution).

Mirrors ``sdk/typescript/src/artifact/nodejs.js``.
"""

from __future__ import annotations

from typing import TYPE_CHECKING

from vorpal_sdk.artifact import Artifact, ArtifactSource
from vorpal_sdk.step import shell
from vorpal_sdk.system import SYSTEMS, get_system_str

if TYPE_CHECKING:
    from vorpal_sdk.context import ConfigContext


class NodeJS:
    """Builder for the Node.js runtime artifact."""

    def build(self, context: ConfigContext) -> str:
        name = "nodejs"
        system = context.get_system()
        system_str = get_system_str(system)

        if system_str == "aarch64-darwin":
            source_target = "darwin-arm64"
        elif system_str == "aarch64-linux":
            source_target = "linux-arm64"
        elif system_str == "x86_64-darwin":
            source_target = "darwin-x64"
        elif system_str == "x86_64-linux":
            source_target = "linux-x64"
        else:
            raise ValueError(f"unsupported {name} system: {system}")

        source_version = "22.22.0"
        source_path = (
            f"https://sdk.vorpal.build/source/"
            f"node-v{source_version}-{source_target}.tar.gz"
        )

        source = ArtifactSource(name, source_path).build()

        step_script = (
            f'cp -pr "./source/{name}/node-v{source_version}-'
            f'{source_target}/." "$VORPAL_OUTPUT"'
        )
        steps = [shell(context, [], [], step_script, [])]

        return (
            Artifact(name, steps, SYSTEMS)
            .with_aliases([f"{name}:{source_version}"])
            .with_sources([source])
            .build(context)
        )
